server/selective_sync_test.go

```diff
@@ -0,0 +1,444 @@
+//go:build exclude
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	goPlugin "github.com/hashicorp/go-plugin"
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/plugin"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func setupPlugin(t *testing.T) *Plugin {
+	t.Helper()
+
+	p := &Plugin{}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	reattachConfigCh := make(chan *goPlugin.ReattachConfig)
+
+	go plugin.ClientMainTesting(ctx, p, reattachConfigCh)
+
+	var reattachConfig *goPlugin.ReattachConfig
+	select {
+	case reattachConfig = <-reattachConfigCh:
+	case <-time.After(5 * time.Second):
+		t.Fatal("failed to get reattach config")
+	}
+
+	socketPath := os.Getenv("MM_LOCALSOCKETPATH")
+	if socketPath == "" {
+		socketPath = model.LocalModeSocketPath
+	}
+
+	clientLocal := model.NewAPIv4SocketClient(socketPath)
+	_, err := clientLocal.ReattachPlugin(ctx, &model.PluginReattachRequest{
+		Manifest:             manifest,
+		PluginReattachConfig: model.NewPluginReattachConfig(reattachConfig),
+	})
+	require.NoError(t, err)
+
+	return p
+}
+
+func setupTeam(t *testing.T, p *Plugin) *model.Team {
+	t.Helper()
+
+	teamName := model.NewRandomTeamName()
+	team, appErr := p.API.CreateTeam(&model.Team{
+		Name:        teamName,
+		DisplayName: teamName,
+		Type:        model.TeamOpen,
+	})
+	require.Nil(t, appErr)
+
+	return team
+}
+func setupUser(t *testing.T, p *Plugin, team *model.Team, isRemote bool) *model.User {
+	t.Helper()
+
+	username := model.NewId()
+	if isRemote {
+		username = fmt.Sprintf("msteams_%s", username)
+	}
+
+	user := &model.User{
+		Email:         fmt.Sprintf("%s@example.com", username),
+		Username:      username,
+		Password:      "password",
+		EmailVerified: true,
+	}
+
+	user, appErr := p.API.CreateUser(user)
+	require.Nil(t, appErr)
+
+	_, appErr = p.API.AddTeamMembers(team.Id, []string{user.Id}, "", false)
+	require.Nil(t, appErr)
+
+	return user
+}
+
+func TestChatSpansPlatforms(t *testing.T) {
+	t.Skip("pending upstream changes: https://mattermost.atlassian.net/browse/MM-57018")
+
+	p := setupPlugin(t)
+
+	t.Run("invalid channel id", func(t *testing.T) {
+		_, appErr := p.ChatSpansPlatforms("")
+		require.Error(t, appErr)
+	})
+
+	t.Run("unknown channel id", func(t *testing.T) {
+		_, appErr := p.ChatSpansPlatforms(model.NewId())
+		require.Error(t, appErr)
+	})
+
+	t.Run("dm between two local users", func(t *testing.T) {
+		team := setupTeam(t, p)
+		user1 := setupUser(t, p, team, false)
+		user2 := setupUser(t, p, team, false)
+
+		channel, err := p.API.CreateDirectChannel(user1.Id, user2.Id)
+		require.Nil(t, err)
+
+		chatSpansPlatforms, appErr := p.ChatSpansPlatforms(channel.Id)
+		require.Nil(t, appErr)
+		assert.False(t, chatSpansPlatforms)
+	})
```

Comment:
This test is "real" -- no mocks, not limited to just the external plugin API. It requires changes in the server to support "reattaching plugins" (see https://mattermost.atlassian.net/browse/MM-57018), but in short this will complement our testcontainers/e2e strategy but for unit tests where mocking remained the only option.